examples/image-resizer/shared/storage.ts

```diff
@@ -0,0 +1,192 @@
+import { S3Client, PutObjectCommand, HeadObjectCommand, GetObjectCommand } from '@aws-sdk/client-s3'
+import { getSignedUrl } from '@aws-sdk/s3-request-presigner'
+
+/**
+ * Storage Adapter Interface
+ * 
+ * TODO: Implement other storage adapters by implementing this interface:
+ * - SupabaseStorageAdapter
+ * - CloudflareR2Adapter  
+ * - GoogleCloudStorageAdapter
+ * - LocalFileSystemAdapter (for development)
+ */
+export interface StorageAdapter {
+  /**
+   * Save a file buffer to storage
+   * @param buffer - File buffer to save
+   * @param filename - Filename/key for the file
+   * @returns Promise resolving to the file URL or key
+   */
+  saveFile(buffer: Buffer, filename: string): Promise<string>
+
+  /**
+   * Get a public or signed URL for a file
+   * @param filename - Filename/key of the file
+   * @returns Public URL or signed URL for the file
+   */
+  getFileUrl(filename: string): string
+
+  /**
+   * Check if a file exists in storage
+   * @param filename - Filename/key to check
+   * @returns Promise resolving to boolean indicating existence
+   */
+  fileExists(filename: string): Promise<boolean>
+
+  /**
+   * Get file buffer (for processing)
+   * @param filename - Filename/key of the file
+   * @returns Promise resolving to file buffer
+   */
+  getFileBuffer(filename: string): Promise<Buffer>
```

Comment:
can it be input stream instead? Buffer uses memory, it's slower and not very scalable, whereas input stream only uses small chunks, because it sends to the filesystem directly.